Generate Mnemonic Wallet

wbobeirne opened the pull request:
Closes #109, and addresses a lot of #557. This PR looks a lot bigger than it really is, so don't let the line count fool ya. This reworks the generate flow to give users the option between a keystore file and a mnemonic phrase. You can see #109 for the discussion around all of the changes, or you can just watch the following example videos:

### [Gfycat - Mnemonic wallet generation](https://gfycat.com/FinishedFlatGartersnake)
### [Gfycat - Updated keystore wallet generation](https://gfycat.com/PlushAromaticHuman)

<br/>

There were a few additional changes related to this:

* Removed generateWallet redux stuff. Everything is handled in state now.
    * Rather than have two separate flows share the same reducer, or add a new reducer, I decided the extremely simple data related to this could be kept in state. If there ever comes a need to output actions, we can add some info-only actions that get fired.
* Removed sidebar from generate altogether.
    * As @tayvano noted, we may want to add back in the guide links somewhere in the flow. Feedback is welcome on where they could fit!
* Re-organized files and styles to better match other tabs.
    * That's where the bulk of the additions / removals come from.

dternyak commented:
Tested, works great!

Implementation looks great as well.

I did notice that there is no way to reset or go back once you have selected a keystore / mnemonic based wallet generation. It would be nice to have a restart / back button, but it doesn't need to hold up the merge.

HenryNguyen5 commented on common/containers/Tabs/GenerateWallet/components/FinalSteps.tsx:
```diff
@@ -0,0 +1,77 @@
+  walletType: WalletType;
+}
+
+export default class FinalSteps extends React.Component<Props> {
```
This could just be an SFC instead, and then you can have all the logic in a separate function too

HenryNguyen5 commented on common/containers/Tabs/GenerateWallet/components/Mnemonic/Mnemonic.tsx:
```diff
@@ -0,0 +1,150 @@
+          </p>
+
+          <div className="GenerateMnemonic-words">
+            {[firstHalf, lastHalf].map((ws, i) => (
```
I would consider unrolling the first level array just to reduce complexity, with having a helper function / SFC generate the `Word` components

wbobeirne replied:
Can you elaborate on
> unrolling the first level array
Not sure what that looks like

wbobeirne replied:
I think I simplified this by making a makeWord helper function.

HenryNguyen5 commented on common/containers/Tabs/GenerateWallet/components/Mnemonic/Mnemonic.tsx:
```diff
@@ -0,0 +1,150 @@
+  };
+
+  private handleConfirmChange = (index: number, value: string) => {
+    console.log(index, value);
```
Got some console logs here

HenryNguyen5 commented on common/containers/Tabs/GenerateWallet/components/Mnemonic/Mnemonic.tsx:
```diff
@@ -0,0 +1,150 @@
+    const confirmValues = [...this.state.confirmValues];
+    confirmValues[index] = value;
+    console.log(confirmValues);
+    this.setState({ confirmValues });
```
This could all be done with a setState cb that'll give you the previous state. Should be more clean.

HenryNguyen5 commented on common/containers/Tabs/GenerateWallet/components/Mnemonic/Word.tsx:
```diff
@@ -0,0 +1,64 @@
+        <div className="MnemonicWord-word input-group">
+          <input
+            className={classnames({
+              'MnemonicWord-word-input': true,
```
This can be shortened to 
```ts
classnames(
               'MnemonicWord-word-input',
               'form-control',
               {'is-valid': word === value}
             )
```

HenryNguyen5 replied:
Or using the same style as LN 49 to keep it consistent

HenryNguyen5 commented on common/containers/Tabs/GenerateWallet/components/Template.tsx:
```diff
@@ -2,23 +2,11 @@ import React from 'react';
-        </div>
-      </div>
-    );
+    return <div className="GenerateWallet Tab-content-pane">{this.props.children}</div>;
```
This could be a SFC instead

HenryNguyen5 commented on common/containers/Tabs/GenerateWallet/components/WalletTypes.tsx:
```diff
@@ -0,0 +1,78 @@
+  onSelect(type: WalletType): void;
+}
+
+export default class WalletTypes extends React.Component<Props, {}> {
```
This could be an SFC too

wbobeirne replied:
Would be dope if there was a linter rule or something that caught implementations of React.Component that only had a render method. I'll convert this, but lmk if you think that's out there.

wbobeirne commented:
Personally I find keeping the markup as together as possible makes reading the structure of the markup a little easier, but I can pretty easily convert some of those maps to helper functions.

wbobeirne commented:
@dternyak added a UI back button, but also refactored this so it no longer uses state to keep track of which wallet we're generating, but uses the router instead. This makes the back button work as well. Added routes for /generate, /generate/keystore, and /generate/mnemonic.

<img width="472" alt="screen shot 2017-12-28 at 2 06 29 pm" src="https://user-images.githubusercontent.com/649992/34420613-8f6fefcc-ebd8-11e7-8b9b-3bcbee4b55ad.png">

@HenryNguyen5 I think I addressed everything! Let me know if I got any changes wrong.

wbobeirne commented:
I alertified the warning text upfront:

<img width="1105" alt="screen shot 2017-12-28 at 2 40 39 pm" src="https://user-images.githubusercontent.com/649992/34421429-31aaa6ca-ebdd-11e7-840c-028d19df07c2.png">
